### PR TITLE
Ensure that we implement expiration timeout to existing cache providers

### PR DESCRIPTION
**Pull request by vaibhav-kaushal**
Ensure that we implement expiration timeout to **existing** caches. 

This is quite useful and lays out the ground-work for implementing Redis as a cache provider.

### CODE REVIEW

File: includes/framework/QAbstractCacheProvider.class.php [hunk 1]
```diff
@@ -13,11 +13,14 @@
...
-		 * @param object $objValue the object to put in the cache
+		 *
+		 * @param string  $strKey        the key to use for the object
+		 * @param object  $objValue      the object to put in the cache
+		 * @param integer $intExpiration Duration for which the cache has to live
+		 *
 		 * @return void
 		 */
-		abstract public function Set($strKey, $objValue);
+		abstract public function Set($strKey, $objValue, $intExpiration);
```
Comment (olegabr) [inline includes/framework/QAbstractCacheProvider.class.php]: make it optional please

Comment (vaibhav-kaushal) [inline includes/framework/QAbstractCacheProvider.class.php]: Done in the latest commit

**Comment by vaibhav-kaushal** [comment]
Sorry about too many commits. Everything is passing now.

**Comment by vaibhav-kaushal** [comment]
@olegabr and @spekary please review and merge. The work on Redis cache provider depends on this. I will not start working on the Redis one unless the base work is in master.

**Comment by vaibhav-kaushal** [comment]
For tests, we will have to make sure that we wait (`sleep(5)`?) and then test whether the data had expired or not. Could you help me at it?

File: includes/framework/QCacheProviderAPC.class.php [hunk 2]
```diff
@@ -21,13 +21,19 @@ public function Get($strKey) {
...
+		 * @param object $objValue              the object to put in the cache
+		 * @param int    $intExpireAfterSeconds Number of seconds after which the object has to expire
+		 *
 		 * @return void
 		 */
-		public function Set($strKey, $objValue) {
-			apc_store ($strKey, $objValue, static::$ttl);
-
+		public function Set($strKey, $objValue, $intExpireAfterSeconds = null) {
+			if($intExpireAfterSeconds) {
```
Comment (vaibhav-kaushal) [inline includes/framework/QCacheProviderAPC.class.php]: Done.

**Comment by vaibhav-kaushal** [comment]
@olegabr Updated the test suite. Please recheck.

**Comment by olegabr** [comment]
Great job, @vaibhav-kaushal!

**Pull request merged**